### PR TITLE
Assorted performance improvements

### PR DESCRIPTION
**Pull request by FunnyMan3595**
## What Does This PR Do
Targeted refactors to improve the performance of some hot methods.

Changes to average total CPU:
* /datum/thrownthing/proc/finalize: -70%
* /atom/movable/CanPass: -64%
* /proc/playsound: -70%
* /turf/Enter: -46%
* /datum/pipeline/Destroy: -97%
* /datum/pipeline/proc/build_pipeline: -54%

Most of these are only hot when lots of items are being thrown around or deleted (e.g. loose singularity), but a couple are pretty significant normally, according to the profile I just took from live:
* /proc/playsound: 58th by total CPU
* /turf/Enter: 93rd by total CPU

## Why It's Good For The Game
Performance go brrr.

## Testing
Turned loose a max-stage singularity, saw nothing (else) go wrong.
Deleted some pipe elements to split networks, got the same pressure but different volumes on each side.

## Changelog
NPFC

### CODE REVIEW

File: code/game/turfs/turf.dm [hunk 1]
```diff
@@ -211,14 +217,15 @@
 		mover.Bump(src, TRUE)
 		return FALSE
 
-	//Finally, check objects/mobs to block entry that are not on the border
+	//Finally, check objects/mobs that block entry and are not on the border
```
Comment (1080pCat) [inline code/game/turfs/turf.dm:220]:
```suggestion
	// Finally, check objects/mobs that block entry and are not on the border
```

File: code/modules/atmospherics/machinery/pipes/pipe.dm [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /obj/machinery/atmospherics/pipe
-	var/datum/gas_mixture/air_temporary //used when reconstructing a pipeline that broke
+	var/datum/gas_mixture/ghost_pipeline //used when reconstructing a pipeline that broke
```
Comment (1080pCat) [inline code/modules/atmospherics/machinery/pipes/pipe.dm:2]:
```suggestion
	var/datum/gas_mixture/ghost_pipeline // used when reconstructing a pipeline that broke
```

File: code/controllers/subsystem/SSthrowing.dm [hunk 3]
```diff
@@ -141,12 +141,9 @@ SUBSYSTEM_DEF(throwing)
...
 	thrownthing.throwing = null
 	if(!hit)
-		for(var/thing in get_turf(thrownthing)) //looking for our target on the turf we land on.
-			var/atom/A = thing
-			if(A == target)
-				hit = 1
-				thrownthing.throw_impact(A, src)
-				break
+		if(get_turf(target) == get_turf(thrownthing))
+			hit = 1
```
Comment (lewcc) [inline code/controllers/subsystem/SSthrowing.dm:145]:
```suggestion
			hit = TRUE
```

File: code/game/atoms_movable.dm [hunk 4]
```diff
@@ -521,9 +521,12 @@
 	return FALSE
 
 /atom/movable/CanPass(atom/movable/mover, turf/target, height=1.5)
-	if(mover in buckled_mobs)
+	// This condition is copied from atom to avoid an extra parent call, because this is a very hot proc.
+	if(!density || !height)
 		return TRUE
-	return ..()
+	if(LAZYIN(buckled_mobs, mover))
```
Comment (lewcc) [inline code/game/atoms_movable.dm:527]: Why not just return lazyin()

File: code/game/turfs/turf.dm [hunk 5]
```diff
@@ -191,16 +191,22 @@
 	// First, make sure it can leave its square
 	if(isturf(mover.loc))
 		// Nothing but border objects stop you from leaving a tile, only one loop is needed
-		for(var/obj/obstacle in mover.loc)
-			if(!obstacle.CheckExit(mover, src) && obstacle != mover && obstacle != forget)
+		// This as anything looks odd, since we check istype shortly after, but it's so we can save an istype check for items, which are far more common than other objects.
+		for(var/obj/obstacle as anything in mover.loc)
```
Comment (lewcc) [inline code/game/turfs/turf.dm:195]: You're continuing on !istype, any reason to do `as anything` in this case?

File: code/game/turfs/turf.dm [hunk 6]
```diff
@@ -191,16 +191,22 @@
...
 				mover.Bump(obstacle, TRUE)
 				return FALSE
 
 	var/list/large_dense = list()
-	//Next, check objects to block entry that are on the border
-	for(var/atom/movable/border_obstacle in src)
+	// Next, check for border obstacles on this turf
+	// Everyting inside a turf is an atom/movable.
+	for(var/atom/movable/border_obstacle as anything in src)
+		if(isitem(border_obstacle) || border_obstacle == forget || isnull(border_obstacle))
```
Comment (lewcc) [inline code/game/turfs/turf.dm:206]: Same here, feels questionable

File: code/game/atoms_movable.dm [hunk 7]
```diff
@@ -521,9 +521,12 @@
...
+		return TRUE
+	return FALSE
```
Comment (FunnyMan3595) [inline code/game/atoms_movable.dm:529]:
```suggestion
	return LAZYIN(buckled_mobs, mover)
```

File: code/game/atoms_movable.dm (hunk 4 above)
Comment (FunnyMan3595) [reply code/game/atoms_movable.dm]: Because I'm a dummy.

File: code/game/turfs/turf.dm (hunk 5 above)
Comment (FunnyMan3595) [reply code/game/turfs/turf.dm:195]: See the comment above.

File: code/game/turfs/turf.dm (hunk 6 above)
Comment (FunnyMan3595) [reply code/game/turfs/turf.dm:206]: `isitem()` is more specific than `atom/movable`

File: code/controllers/subsystem/SSthrowing.dm [hunk 8]
```diff
@@ -141,12 +141,9 @@ SUBSYSTEM_DEF(throwing)
 	//done throwing, either because it hit something or it finished moving
 	thrownthing.throwing = null
 	if(!hit)
-		for(var/thing in get_turf(thrownthing)) //looking for our target on the turf we land on.
-			var/atom/A = thing
-			if(A == target)
-				hit = 1
-				thrownthing.throw_impact(A, src)
-				break
```
Comment (DGamerL) [inline code/controllers/subsystem/SSthrowing.dm:149]: This was such a silly way of doing this

**Review by DGamerL — APPROVED** [review]
Please deconflict for merge

**Review by DGamerL — CHANGES_REQUESTED** [review]
Right this also needs a TM

File: code/controllers/subsystem/SSthrowing.dm (hunk 8 above)
Comment (FunnyMan3595) [reply code/controllers/subsystem/SSthrowing.dm]:
```dm
for(var/word in GLOB.words)
	if(word == "yeah")
		to_chat(user, "[capitalize(word)]...")
		break
```

**Comment by FunnyMan3595** [comment]
> Please deconflict for merge

Done (finally)

**Review by hal9000PR — APPROVED** [review]
LGTM

**Pull request merged**